### PR TITLE
Less allocs during `flatMapErrorThrowing` and `flatMapThrowing`

### PR DESCRIPTION
**Pull request by fabianfett**
Less allocs = Faster code

### Motivation:

- We alloc quite a lot with our implementations of `flatMapThrowing` and `flatMapErrorThrowing`.
- While we don't use Futures a lot in NIO itself a lot of our users, depend quite a bit on them. Let's make their code faster.

### Modifications:

- Create a `Promise` and use `_whenComplete` directly instead of going through another `flatMap` method

### Result:

- In my testing I see a reduction of 3 allocs per invocation 🎉

### CODE REVIEW

**Review by Lukasa — APPROVED** [review]
Cool, LGTM. Do we have allocation counter tests for these? If not, it'd be nice to have them.

File: Sources/NIO/EventLoopFuture.swift [hunk 1]
```diff
@@ -506,13 +506,22 @@ extension EventLoopFuture {
...
-                return EventLoopFuture<NewValue>(eventLoop: self.eventLoop, error: error, file: file, line: line)
+        let next = EventLoopPromise<NewValue>(eventLoop: eventLoop, file: file, line: line)
+        self._whenComplete {
+            switch self._value! {
+            case .success(let t):
+                do {
+                    let r = try callback(t)
+                    return next._setValue(value: .success(r))
+                }
+                catch {
```
Comment (glbrntt) [inline Sources/NIO/EventLoopFuture.swift:517]: nit

```suggestion
                } catch {
```

File: Sources/NIO/EventLoopFuture.swift [hunk 2]
```diff
@@ -531,13 +540,22 @@ extension EventLoopFuture {
...
+        self._whenComplete {
+            switch self._value! {
+            case .success(let t):
+                return next._setValue(value: .success(t))
+            case .failure(let e):
+                do {
+                    let r = try callback(e)
+                    return next._setValue(value: .success(r))
+                }
+                catch {
```
Comment (glbrntt) [inline Sources/NIO/EventLoopFuture.swift:553]:
```suggestion
                } catch {
```

**Review by glbrntt — APPROVED** [review]
Nice!

**Review by weissi — APPROVED** [review]
🎉 thank you!

**Pull request merged**